extracting APRs module

gmbronco opened the pull request.

gmbronco committed:
extracting APRs module

franzns committed:
Merge branch 'v3-canary' into aprs-module

franzns commented on modules/aprs/apr-repository.ts:
```diff
@@ -0,0 +1,128 @@
+     * Only updates when the APR value has changed
+     * @returns changed poolIDs
+     */
+    async savePoolAprItems(aprItems: Omit<PrismaPoolAprItem, 'createdAt' | 'updatedAt'>[]): Promise<string[]> {
```
Could be simplified by passing a chain argument. The handlers are always chain specific that use this.

gmbronco replied:
Yes, let's do that per chain

franzns commented on modules/aprs/handlers/swap-fee-apr/swap-fee-apr-handler.ts:
```diff
@@ -0,0 +1,58 @@
+
+                let protocolFee = parseFloat(pool.dynamicData.protocolSwapFee);
+                if (pool.type === 'GYROE') {
+                    protocolFee = parseFloat(pool.dynamicData.protocolYieldFee || '0');
```
This seems odd. Need to double check how we sync gyro protocol fees.

franzns commented on modules/aprs/apr-repository.ts:
```diff
@@ -0,0 +1,128 @@
+
+export type PoolAPRData = Prisma.PrismaPoolGetPayload<typeof aprsInclude>;
+
+export class AprRepository {
```
I find the name confusing. Would call it something like `AprDatabaseManager` or `AprDataManager`

franzns committed:
fix yield fee collection bool

franzns commented on modules/aprs/apr-manager.ts:
```diff
@@ -0,0 +1,68 @@
+        }
+
+        if (failedCalculators.length > 0) {
+            console.warn(`The following APR calculators failed: ${failedCalculators.join(', ')}`);
```
If this happens, I think we still want some alert in sentry

franzns commented on modules/aprs/types.ts:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Types of APR calculators available in the system
+ */
+export type AprCalculatorType =
+    | 'swapFeeApr'
+    | 'boostedPoolApr'
+    | 'dynamicSwapFeeApr'
+    | 'gaugeApr'
+    | 'veBalProtocolApr'
+    | 'veBalVotingApr'
+    | 'morphoRewardsApr'
+    | 'ybTokensApr'
+    | 'aaveApiApr'
+    | 'masterchefFarmApr'
+    | 'reliquaryFarmApr'
+    | 'beetswarsGaugeVotingApr'
+    | 'quantAmmApr';
+
+/**
+ * Configuration for APR calculators
+ */
+export interface AprCalculatorConfig {
+    type: AprCalculatorType;
+    params?: Record<string, any>;
+}
```
`AprCalculatorType` and `AprCalculatorConfig` seems unused

franzns commented on modules/aprs/README.md:
```diff
@@ -0,0 +1,98 @@
+The module follows a layered architecture:
+
+1. **Repository Layer** - Handles data access (`PoolAprRepository`)
+2. **Handlers Layer** - Contains pure calculation logic (`AprHandler` implementations)
+3. **Configuration Layer** - Defines which calculators to use for each chain
+4. **Manager Layer** - Coordinates calculators and persistence (`AprManager`)
```
Either call them Handlers or Calculators, its mixed currently.

franzns commented on modules/aprs/handlers/yb-tokens/yb-tokens-apr-handler.ts:
```diff
@@ -0,0 +1,146 @@
+        private aprConfig: YbAprConfig,
+        chain: Chain,
+    ) {
+        this.ybTokensAprHandlers = new YbAprHandlers(this.aprConfig, chain);
```
Wonder if we need to dissect the `YbAprHandlers` a bit as well. In a first run, I think we could copy them as-is and refactor them in a second phase.

franzns committed:
move config

franzns committed:
add dynamic swapfee apr handler

franzns committed:
add mabeets apr

franzns committed:
move types

franzns committed:
add morpho rewards handler

franzns committed:
add nested pool apr service

franzns committed:
add quant amm apr handler

franzns committed:
liquidity gauge apr handler

franzns committed:
rename calculator to handler

franzns committed:
add vebal apr handlers

franzns committed:
remove config from aave rewards and enable it on chains

franzns committed:
remove items that dont exist anymore

franzns committed:
add merkl

franzns committed:
add but dont enable 7d30d swap fee apr handler

franzns committed:
fix imports

franzns committed:
sync incentivized category after apr sync

franzns committed:
replace old service with new one

franzns committed:
add surplus handler, move dynamic fee handler

franzns committed:
integrate surplus apr

franzns committed:
integrate apr items, get rid of apr group

franzns committed:
get rid of apr range

franzns committed:
types

franzns committed:
cleanup and move yb apr handlers

franzns committed:
drop apr range

franzns committed:
fix merkl query

franzns committed:
fix await

franzns committed:
fix update total apr

franzns committed:
fix beetswars

franzns committed:
fux yusd apr

franzns committed:
fix yusd2 apr base

franzns committed:
changeset

franzns committed:
Merge branch 'v3-canary' into aprs-module

franzns committed:
add back aave for avax

gmbronco commented:
MaBeetsAprHandler updates prismaPoolStakingReliquaryFarmLevel, ideally it would be a separate action

gmbronco commented:
Overall looks great, most of the services are returning APRs without any DB access. Few remarks:

* use DI for the tokenService - I updated the LiquidityGaugeAprHandler and added an interface, so it's easy to pass in the constructors, but some handlers are still calling the DB directly, eg: quant amm
* mix in side effects in the controllers, outside of the module, so it's easier to test and compose
* nice to have: prismaPoolStakingReliquaryFarmLevel updates are mixed into the handler, possible to extract? your call.

franzns committed:
move logic to controller

franzns committed:
extract tokenservice from handlers for DI

franzns committed:
add stability apr

franzns committed:
Merge branch 'v3-canary' into aprs-module
